common/cuda_hip/reorder/rcm_kernels.hpp.inc

```diff
@@ -0,0 +1,608 @@
+    // find beginning of all components
+    auto it = thrust::make_counting_iterator(size_type{});
+    const auto predicate = adj_not_predicate<IndexType>{sorted_component_ids};
+    const auto num_components = static_cast<size_type>(
+        thrust::count_if(policy, it, it + num_rows, predicate));
+    result.set_num_components(num_components);
+    const auto ptrs = result.ptrs.get_data();
+    const auto min_deg_node = result.min_deg_node.get_data();
+    thrust::copy_if(policy, it, it + num_rows, ptrs, predicate);
+    // set the sentinel entry
+    set_element(result.ptrs, num_components, num_rows);
+    // find minimum degree node for each component
+    array<IndexType> component_id_array{exec, num_components};
+    const auto component_ids = component_id_array.get_data();
+    thrust::reduce_by_key(
+        policy, sorted_component_ids, sorted_component_ids + num_rows, nodes,
+        component_ids, min_deg_node, thrust::equal_to<IndexType>{},
+        node_min_degree_reduction<IndexType>{num_rows, node_degrees});
+    // map component IDs to consecutive indexing
+    array<IndexType> compacted_node_component{exec,
+                                              static_cast<size_type>(num_rows)};
+    thrust::lower_bound(policy, component_ids, component_ids + num_components,
+                        node_component, node_component + num_rows,
+                        compacted_node_component.get_data());
+    result.node_component = std::move(compacted_node_component);
+    return result;
+}
+
+
+template <typename IndexType>
+struct ubfs_levels {
+    /** Mapping node -> level */
+    array<IndexType> node_level;
+    /** Segmented list of nodes for each level */
+    array<IndexType> nodes;
+    /** Pointers into nodes for each level */
+    array<IndexType> ptrs;
+    /** How many levels are there? */
+    IndexType num_levels;
+
+    ubfs_levels(std::shared_ptr<const DefaultExecutor> exec, size_type num_rows)
+        : node_level{exec, num_rows},
+          nodes{exec, num_rows},
+          ptrs{exec, num_rows + 1}
+    {}
+};
+
+
+template <typename IndexType>
+__global__ __launch_bounds__(default_block_size) void ubfs_level_kernel(
+    const IndexType* __restrict__ row_ptrs,
+    const IndexType* __restrict__ col_idxs,
+    const IndexType* __restrict__ sources, size_type num_sources,
+    IndexType level, IndexType* __restrict__ node_levels,
+    IndexType* __restrict__ level_nodes, IndexType* __restrict__ output_ptr)
+{
+    using atomic_type = typename atomic_map<IndexType>::type;
+    const auto source = thread::get_thread_id_flat();
+    if (source >= num_sources) {
+        return;
+    }
+    const auto row = sources[source];
+    const auto begin = row_ptrs[row];
+    const auto end = row_ptrs[row + 1];
+    atomic_type unsigned_unattached{};
+    const auto unattached = invalid_index<IndexType>();
+    memcpy(&unsigned_unattached, &unattached, sizeof(IndexType));
```

Comment:
I think casting a value to a type that cannot represent it is UB, but I'm not 100% on that. Soon, I will add `atomic_cas` functions with signed integers myself that work without this `memcpy` approach with inline assembly.

Reply:
From cppreference, the conversion signed to unsigned integer is _implementation_ defined (at least until c++20). So it should be valid here. But it's not important.

Reply:
For the CAS to work correctly, I need them to be bitwise identical, until two's complement is standardized, and because this is a temporary solution until we have nice well-typed atomics, I would like to leave the safe version.